go-runtime/encoding/encoding.go

```diff
@@ -31,37 +38,34 @@ func encodeValue(v reflect.Value, w *bytes.Buffer) error {
 		w.WriteString("null")
 		return nil
 	}
+
 	t := v.Type()
-	switch {
-	case t.Kind() == reflect.Ptr && t.Elem().Implements(jsonMarshaler):
-		v = v.Elem()
-		fallthrough
 
-	case t.Implements(jsonMarshaler):
-		enc := v.Interface().(json.Marshaler) //nolint:forcetypeassert
-		data, err := enc.MarshalJSON()
+	// Special-cased types
+	switch {
+	case t == reflect.TypeFor[time.Time]():
+		data, err := json.Marshal(v.Interface())
 		if err != nil {
 			return err
 		}
 		w.Write(data)
 		return nil
 
-	case t.Kind() == reflect.Ptr && t.Elem().Implements(textMarshaler):
-		v = v.Elem()
-		fallthrough
-
-	case t.Implements(textMarshaler):
-		enc := v.Interface().(encoding.TextMarshaler) //nolint:forcetypeassert
-		data, err := enc.MarshalText()
-		if err != nil {
-			return err
-		}
-		data, err = json.Marshal(string(data))
+	case t == reflect.TypeFor[json.RawMessage]():
```

Comment:
I don't think we're using `json.RawMessage` for anything important, just for slightly more type safety. So if you want to replace that usage you mentioned with `[]byte`, that's fine too.